Multi stage doc improvements (#332)
pinot-contrib/pinot-docs: users/user-guide-query/multi-stage-query/operator-types/union.md

---
description: >-
  Describes the union relation operator in the multi-stage query engine.
---

# Union relational operator

The union operator combines the results of two or more queries into a single result set.
The result set contains all the rows from the queries.

> Contrary to other set operations ([intersect](intersect.md) and [minus](minus.md)), the union operator does not remove

Reviewer: this confused me a bit.. although the operator doesnt remove duplicates, we have a way of achieving union by adding the aggregate. So from end user perspective the functionality works. Reading this made me first think we don't support the functionality

Author: I don't get what do you mean. Are you suggesting that the user can add a `distinct` or an `aggregate` phase after the union so they can achieve the same thing?

Author: Oh, I guess you mean that we actually honor the fact that `all` is provided or not. Yes, the text is pretty confusing.

duplicates from the result set.
Therefore its semantic is similar to the SQL `UNION` or `UNION ALL` operator.

There is no guarantee on the order of the rows in the result set.

{% hint style="info" %}

> While `EXCEPT` and `INTERSECT` SQL clauses does not support the `ALL` modifier, the `UNION` clause does.

Reviewer 2: ```suggestion
While `EXCEPT` and `INTERSECT` SQL clauses do not support the `ALL` modifier, the `UNION` clause does.
```

{% endhint %}

## Implementation details
The current implementation consumes input relations one by one.
It first returns all rows from the first input relation, then all rows from the second input relation, and so on.

### Blocking nature
The union operator is a streaming operator that consumes the input relations one by one.
The current implementation fully consumes the inputs in order.
See [the order of input relations matter](#the-order-of-input-relations-matter) for more details.

## Hints
None

## Stats
### executionTimeMs
Type: Long

The summation of time spent by all threads executing the operator.
This means that the wall time spent in the operation may be smaller that this value if the parallelism is larger than 1.

### emittedRows
Type: Long

The number of groups emitted by the operator.

## Explain attributes

The union operator is represented in the explain plan as a `LogicalUnion` explain node.

### all
Type: Boolean

Whether the union operator should remove duplicates from the result set.

> Although Pinot the SQL `UNION` and `UNION ALL` _clauses_ are supported, the union _operator_ does only support the 

Reviewer 2: ```suggestion
Although Pinot supports the SQL `UNION` and `UNION ALL` _clauses_, the union _operator_ does only support the 
```

`UNION ALL` semantic.
In order to implement the `UNION` semantic, the multi-stage query engine adds an extra [aggregate](./aggregate.md) to 

> calculate the _distinct_.

Reviewer 2: Incomplete sentence?

For example the plan of:
```sql
select userUUID
from (select userUUID from userAttributes)
UNION ALL
(select userUUID from userGroups)
```

> Is the expected:

Reviewer 2: ```suggestion
Is expected to be:
```

      LogicalTableScan(table=[[default, userAttributes]])
  PinotLogicalExchange(distribution=[hash[0]])
    LogicalProject(userUUID=[$4])
      LogicalTableScan(table=[[default, userGroups]])
```

While the plan of:
```sql
explain plan for
select userUUID
from (select userUUID from userAttributes)
UNION -- without ALL!
(select userUUID from userGroups)
```

Is a bit more complex
```
LogicalAggregate(group=[{0}])
  PinotLogicalExchange(distribution=[hash[0]])
    LogicalAggregate(group=[{0}])
      LogicalUnion(all=[true])
        PinotLogicalExchange(distribution=[hash[0]])
          LogicalProject(userUUID=[$6])
            LogicalTableScan(table=[[default, userAttributes]])
        PinotLogicalExchange(distribution=[hash[0]])
          LogicalProject(userUUID=[$4])
            LogicalTableScan(table=[[default, userGroups]])
```

Notice that `LogicalUnion` is still using `all=[true]` but the `LogicalAggregate` is used to remove the duplicates.
This also means that while the union _operator_ is always streaming, the union _clause_ results in a blocking plan
(given the [aggregate](./aggregate.md#blocking-nature) operator is blocking).

## Tips and tricks

> ### The order of input relations matter

Reviewer 2: ```suggestion
### The order of input relations matters
```
